types/src/trusted_state.rs

```diff
@@ -94,17 +94,27 @@ impl TrustedState {
                     )
                 })?;
 
-            // Verify the latest ledger info inside the latest epoch.
+            // If the latest ledger info is in the same epoch as the new verifier, verify it and
+            // use it as latest state, otherwise fallback to the epoch change ledger info.
+            let new_epoch = new_epoch_state.epoch;
             let new_verifier = Arc::new(new_epoch_state);
 
-            // If these are the same, then we do not have a LI for the next Epoch and hence there
-            // is nothing to verify.
-            if epoch_change_li != latest_li {
-                new_verifier.verify(latest_li)?;
-            }
+            let verified_state =
+                Waypoint::new_any(if new_epoch == latest_li.ledger_info().epoch() {
+                    new_verifier.verify(latest_li)?;
+                    latest_li.ledger_info()
+                } else {
+                    ensure!(
+                        latest_li == epoch_change_li
+                            || latest_li.ledger_info().epoch() > new_epoch
+                                && epoch_change_proof.more,
+                        "Invalid epoch in latest ledger info",
+                    );
+                    epoch_change_li.ledger_info()
+                });
```

Comment:
this isn't really readable.... the level of nesting here is :trollface:

Reply:
after reading this a dozen times I can see the logic is correct, but we should really break it up with a clear comment that you're moving into a partial state as you attempt to obtain a new set of li.

There are actually three states here...

the new li doesn't move us into a new epoch, okay, let' s just use it
the new li is in the new epoch, so consume the epoch change li
the new li is beyond the new epoch but there are more epochs to come, so consume the epoch change li
then we have the one error case, which is new li is beyond the the new epoch but we don't have more epochs coming, error

can we write this as a if else if else if else  that returns the appropriate li?

Reply:
It looks like the key change here is `latest_li.ledger_info().epoch() > new_epoch && epoch_change_proof.more`, i.e., there was a gap and you need more epoch change li's to catch up, so just ratchet as far as you can. Actually, this makes me wonder why we even have `more`? Could just be an implicit property like "there is a gap" ==> "more".

Reading through how `libra-client` uses this also reminded me how we're not checking other proofs lol. 😭

Reply:
agree `more` is more like a hint. I don't know what's our story of verifying-api overall.